### PR TITLE
hase: init at unstable-2020-10-06

### PR DESCRIPTION
**Pull request by uninsane**
project page: <http://ziz.gp2x.de/hase/index.htm>

"Hase is an open source (GPL) gravity based artillery shooter. It is similar to Worms, Hedgewars or artillery, but the gravity force and direction depends on the mass nearby."

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    repo = "sparrow3d";
+    rev = "2033349d7adeba34bda2c442e1fec22377471134";
+    hash = "sha256-28j5nbTYBrMN8BQ6XrTlO1D8Viw+RiT3MAl99BAbhR4=";
+  };
+
+  nativeBuildInputs = [
+    pkg-config
+  ];
+
+  propagatedBuildInputs = [
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:29]:
```suggestion
  buildInputs = [
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    SDL.dev
+    SDL_image
+    SDL_ttf
+    SDL_mixer
+    SDL_net
+  ];
+
+  postConfigure = ''
+    # sparrow expects the SDL_* libraries to already be on the ld path
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:38]:
```suggestion
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    NIX_CFLAGS_COMPILE=$(pkg-config --cflags SDL_image SDL_ttf SDL_mixer SDL_net)
+  '';
+
+  buildPhase = ''
+    make dynamic
+  '';
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:44]:
```suggestion
  buildFlags = [ "dynamic" ];
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  installPhase = ''
+    # `dev` output
+    mkdir -p $dev/{include,lib/pkgconfig}
+    cp sparrow*.h $dev/include
+    substituteAll ${./sparrow3d.pc.in} $dev/lib/pkgconfig/sparrow3d.pc
+
+    # `out` output
+    mkdir -p $out/lib
+    cp libsparrow{3d,Net,Sound}.so $out/lib
+  '';
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:55]:
```suggestion
  installPhase = ''
    runHook preInstall

    mkdir -p $out/lib
    cp libsparrow{3d,Net,Sound}.so $out/lib

    mkdir -p $dev/{include,lib/pkgconfig}
    cp sparrow*.h $dev/include
    substituteAll ${./sparrow3d.pc.in} $dev/lib/pkgconfig/sparrow3d.pc

    runHook postInstall
  '';
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  checkPhase = ''
+    make all_no_static
+    ./testfile.sh
+    # the other test cases are interactive
+    # ./testsparrow
+  '';
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:62]:
```suggestion
  doCheck = true;
  
  checkPhase = ''
    runHook preCheck
  
    make all_no_static
    ./testfile.sh
    
    runHook postCheck  
  '';
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  doCheck = true;
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:64]:
```suggestion
```

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+  meta = with lib; {
+    description = "a software renderer for different open handhelds like the gp2x, wiz, caanoo and pandora";
+    homepage = "https://github.com/theZiz/sparrow3d";
+    license = licenses.lgpl21;
+    maintainers = with maintainers; [ colinsane ];
+    platforms = platforms.linux;
+  };
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:72]:
```suggestion
  meta = {
    homepage = "https://github.com/theZiz/sparrow3d";
    description = "A software renderer for different open handhelds like the gp2x, wiz, caanoo and pandora";
    license = lib.licenses.lgpl21;
    maintainers = with lib.maintainers; [ colinsane ];
    platforms = platforms.linux;
  };
```

File: pkgs/development/libraries/sparrow3d/sparrow3d.pc.in [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+Version: @version@
+Requires: \
+  sdl \
+  SDL_image \
+  SDL_ttf \
+  SDL_mixer \
+  SDL_net
+Cflags: -isystem${includedir}
+Libs: -L${libdir} -lsparrow3d -lsparrowNet -lsparrowSound
+
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/sparrow3d.pc.in:17]:
```suggestion
```

File: pkgs/games/hase/default.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  pname = "hase";
+  version = "unstable-2020-10-06";
+
+  src = fetchFromGitHub {
+    owner = "theZiz";
+    repo = "hase";
+    rev = "31d6840cdf0c72fc459f10402dae7726096b2974";
+    hash = "sha256-d9So3E8nCQJ1/BdlwMkGbaFPT9mkX1VzlDGKp71ptEE=";
+  };
+  patches = [ ./prefer-dynamic.patch ];
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:19]:
```suggestion

  patches = [ ./prefer-dynamic.patch ];
```

File: pkgs/games/hase/default.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+    pkg-config
+  ];
+
+  buildInputs = [
+    sparrow3d
+    zlib
+  ];
+
+  buildPhase = ''
+    # hase expects its dependencies to already be on the ld path
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:31]:
```suggestion
```

File: pkgs/games/hase/default.nix [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    ./install.sh $out
+  '';
+
+  postFixup = ''
+    substituteInPlace "$out/share/applications/hase.desktop" \
+      --replace "Exec=hase" "Exec=$out/bin/hase"
+  '';
+
+  meta = with lib; {
+    description = "Hase is an open source gravity based artillery shooter. It is similar to Worms, Hedgewars or artillery, but the gravity force and direction depends on the mass nearby. It is optimized for mobile game consoles like the GP2X, Open Pandora or GCW Zero";
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:45]:
```suggestion
    description = "An open-source artillery shooter";
    longDescription = ''
      Hase is an open source gravity based artillery shooter. It is similar to
      Worms, Hedgewars or artillery, but the gravity force and direction
      depends on the mass nearby. It is optimized for mobile game consoles like 
      the GP2X, Open Pandora or GCW Zero.
    '';
```

File: pkgs/games/hase/default.nix [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+    mkdir -p $out/{bin,share/applications,share/pixmaps}
+    ./install.sh $out
+  '';
+
+  postFixup = ''
+    substituteInPlace "$out/share/applications/hase.desktop" \
+      --replace "Exec=hase" "Exec=$out/bin/hase"
+  '';
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:44]:
```suggestion
  meta = {
```

File: pkgs/games/hase/default.nix [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    description = "Hase is an open source gravity based artillery shooter. It is similar to Worms, Hedgewars or artillery, but the gravity force and direction depends on the mass nearby. It is optimized for mobile game consoles like the GP2X, Open Pandora or GCW Zero";
+    homepage = "http://ziz.gp2x.de/hase/";
+    license = licenses.gpl3;
+    maintainers = with maintainers; [ colinsane ];
+    platforms = platforms.linux;
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:49]:
```suggestion
    license = lib.licenses.gpl3Plus;
    maintainers = with lib.maintainers; [ colinsane ];
    platforms = lib.platforms.linux;
```

File: pkgs/games/hase/default.nix [hunk 14]
```diff
@@ -0,0 +1,51 @@
+{ lib
+, fetchFromGitHub
+, pkg-config
+, stdenv
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:4]:
```suggestion
```

File: pkgs/games/hase/default.nix [hunk 15]
```diff
@@ -0,0 +1,51 @@
+{ lib
+, fetchFromGitHub
```
Comment (AndersonTorres) [inline pkgs/games/hase/default.nix:2]:
```suggestion
, stdenv
, fetchFromGitHub
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/sparrow3d/default.nix (hunk 1 above)
Comment (uninsane) [reply pkgs/development/libraries/sparrow3d/default.nix:29]: changing `propagatedBuildInputs` to `buildInputs` here causes `nix buld '.#hase'` to fail with:
```
@nix { "action": "setPhase", "phase": "buildPhase" }
building
Package sdl was not found in the pkg-config search path.
Perhaps you should add the directory containing `sdl.pc'
to the PKG_CONFIG_PATH environment variable
Package 'sdl', required by 'sparrow3d', not found
```

sparrow both links to sdl and when a project like Hase `#include <sparrow3d.h>`s that in turn `#include <SDL.h>`s. i think `propagatedBuildInputs` is the solution here, though maybe the sparrow3d.pc pkg-config file could `-Isystem ${sdl.dev}/include/SDL` & friends as an alternative, if you think that's better?

File: pkgs/games/hase/default.nix (hunk 13 above)
Comment (uninsane) [reply pkgs/games/hase/default.nix:55]: for the license (gpl3 -> gpl3Plus): i was going off <https://github.com/theZiz/hase/blob/master/LICENSE>.

it doesn't make any mention of gpl3Only v.s. gpl3Plus, and i'm not knowledgeable enough to know which one is appropriate. i'll apply this suggestion though -- i guess when a GPL3 license doesn't explicitly say otherwise, GPL3+ is assumed?

**Comment by uninsane** [comment]
@AndersonTorres feedback applied, except for changing `propagatedBuildInputs` to `buildInputs` as that breaks the Hase build.

File: pkgs/development/libraries/sparrow3d/default.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/libraries/sparrow3d/default.nix:29]: Well, two things:

1. I believe you can use the function `makePkgconfigItem` to generate pkg-config `.pc` files;
2. I don't know how other distros handle things like this. _However_, given that other distros just throw everything in a global namespace, they are not useful to draw a line here.

OK, keep `propagated???`

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/sparrow3d/default.nix (hunk 1 above)
Comment (uninsane) [reply pkgs/development/libraries/sparrow3d/default.nix:54]: switched to `makePkgconfigItem` (thanks for letting me know about it!) & kept `propagatedBuildInputs`

File: pkgs/top-level/all-packages.nix [hunk 16]
```diff
@@ -23552,6 +23552,8 @@ with pkgs;
 
   spaceship-prompt = callPackage ../shells/zsh/spaceship-prompt {};
 
+  sparrow3d = callPackage ../development/libraries/sparrow3d {};
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:23555]: Split sparrow3d into its own commit (no need to open another PR).

In case of doubts, use this tutorial:
https://stackoverflow.com/questions/6217156/break-a-previous-commit-into-multiple-commits

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 16 above)
Comment (uninsane) [reply pkgs/top-level/all-packages.nix]: okay, i've split the commit

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/development/libraries/sparrow3d/default.nix [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+
+  checkPhase = ''
+    runHook preCheck
+
+    make all_no_static
+    ./testfile.sh
+
+    runHook postCheck
+  '';
+
```
Comment (AndersonTorres) [inline pkgs/development/libraries/sparrow3d/default.nix:90]:
```suggestion
```

**Review by AndersonTorres — APPROVED** [review]
Besides LGTM
